src/application_settings/container_section_base.py

```diff
@@ -85,12 +91,67 @@ def _set(self) -> Self:
             subsec._set()  # pylint: disable=protected-access
         return self
 
+    @staticmethod
+    def _is_parameter_container_section(a_field_type: Any) -> bool:
+        try:
+            return issubclass(a_field_type, ParameterContainerSectionProtocol)
+        except TypeError:
+            return False
+
+    def _check_initialized_and_extra(
+        self, data: dict[str, Any], section_name: str = ""
+    ) -> Self:
+        """Store the singleton and check if extra parameters were provided and/or parameters were missing."""
+        if is_pydantic_dataclass(type(self)):
+            section_specifier = (
+                f" in section {section_name}"
+                if section_name
+                else " in the root section"
+            )
+            field_names = {fld.name for fld in fields(self)}  # type: ignore[arg-type]
+            subsection_names = {
+                fld.name
+                for fld in fields(self)  # type: ignore[arg-type]
+                if self._is_parameter_container_section(fld.type)
+            }
+            uninitialized_field_names = (
+                field_names - subsection_names - set(data.keys())
+            )
+            for uninitialized_field in uninitialized_field_names:
+                logger.log(
+                    log_level(self.kind()),
+                    f"Parameter {uninitialized_field}{section_specifier} initialized with default value.",
+                )
+            extra_data_fields = set(data.keys()) - field_names
+            for extra_data_field in extra_data_fields:
+                logger.log(
+                    log_level(self.kind()),
+                    f"Extra parameter {extra_data_field}{section_specifier} that is not used for initialization.",
+                )
+            subsections = [
+                (
+                    subsection_name,
+                    cast(
+                        ParameterContainerSectionProtocol,
+                        getattr(self, subsection_name),
+                    ),
+                )
+                for subsection_name in subsection_names
+            ]
+            prefix = f"{section_name}." if section_name else ""
+            for subsec_name, subsec in subsections:
+                sub_data = data.get(subsec_name, {})
+                subsec._check_initialized_and_extra(  # pylint: disable=protected-access
+                    sub_data, f"{prefix}{subsec_name}"
+                )
```

Comment:
These warnings are really nice!

the library is becoming quite hard to follow though. probably also because i'm not putting in the work :-).

Reply:
I added a bunch of comments, hopefully that makes it easier to understand.

Would like to know which other parts you find hard to follow...